### DISCUSSION TITLE
draw numbers on TFT has sometimes ghost-letters on the beginn and/or the end of the line

### DISCUSSION THREAD

**Discussion by Loetnase** [Bodmer/TFT_eSPI#3459] [Q&A - Fonts]
Look for the picture over the red marks.
![20240824_144045_2](https://github.com/user-attachments/assets/3a3392b5-3f47-4c9f-8e72-83629d16372e)
I have testet different types of draw, all have the same problem,
i would like use the three drawRightString, drawCenterString and drawString.
But i got this ghost-letters on the beginn and/or the end of the line.   
This .ino is running on a ESP32 with ILI9341 320x240, TFT_eSPI is the newest 2.5.43
The Arduino IDE is the new 2.3.2
Do i make there a mistake or can that be a bug in the TFT_eSPI.

In the next post is the Software with ther you can see my problem live.

Thank you for your help.

**Comment by Loetnase** [comment]
// https://github.com/Bodmer/TFT_eSPI
// This .ino is running on a ESP32 with ILI9341 320x240 
// TFT_eSPI is the newest 2.5.43

#include <TFT_eSPI.h>  //  Hardware Bibliothek von Bodmer mit ILI9341 mit 240 x 320 Pixel

TFT_eSPI tft = TFT_eSPI();
float value = 0;
float d = 270;

void setup() {

  // backlight for the TFT on my System GPIO15, your System maybe use an differen GPIO
  pinMode(15, OUTPUT);
  digitalWrite(15, LOW);

  tft.init();
  tft.setRotation(0);  // rotation set 0-3 = 0-270 Grad
  tft.fillScreen(TFT_RED);
}

void loop() {
  // Create a sine wave for testing
  d += 0.9;  // 0.1 = speed of the sine
  if (d >= 360) d = 0;
  value = 10250 * sin((d + 0) * 0.0174532925);

  tft.setTextColor(TFT_WHITE, TFT_BLACK, 0);  //TFT Text Color WHITE on BLACK
  char buf[8];
  dtostrf(value, 5, 0, buf);  // float value to string buf

  tft.drawString(buf, 60, 5, 1);
  tft.drawCentreString(buf, 120, 5, 1);
  tft.drawRightString(buf, 180, 5, 1);
  tft.drawCentreString("drawString, CentreString, RightString", 120, 20, 2);
  tft.drawString(buf, 40, 40, 2);
  tft.drawCentreString(buf, 120, 40, 2);
  tft.drawRightString(buf, 200, 40, 2);
  tft.drawCentreString("tft.drawString(buf)", 120, 60, 2);
  tft.drawString(buf, 120, 80, 4);
  tft.drawCentreString("drawCentreString(buf)", 120, 110, 2);
  tft.drawCentreString(buf, 120, 130, 4);
  tft.drawCentreString("tft.drawRightString", 120, 160, 2);
  tft.drawRightString(buf, 120, 180, 4);
  tft.drawCentreString("tft.drawFloat(value)", 120, 210, 2);
  tft.drawFloat(value, 2, 70, 230, 4);
  tft.drawCentreString("tft.drawNumber(value)", 120, 260, 2);
  tft.drawNumber(value, 80, 280, 4);
}

**Reply by MYCAMEL222** [reply]
Add tft.fillScreen(TFT_RED); at the beginning of the loop before "// Create a sine wave for testing". That will "clear" the screen of the last characters. You can expect flicker between redraws. Maybe add a delay(1000); at the end of the loop.

The best approach is to draw numbers to sprites and push them to the red screen for no flicker....
